### PR TITLE
[text analytics] add string-index-type support

### LINKED ISSUE #12032: [text analytics] add support for optional string-index-type query param

### PR DESCRIPTION
**Pull request by iscai-msft**
Defaults to Uniode code points for Python

Following talks with @johanste python is going to at least start out not exposing the string-index-type param to users

fixes #12032

### CODE REVIEW

File: sdk/textanalytics/azure-ai-textanalytics/tests/test_analyze_sentiment.py [hunk 1]
```diff
@@ -16,7 +16,7 @@
     TextAnalyticsClient,
     TextDocumentInput,
     VERSION,
-    TextAnalyticsApiVersion
+    TextAnalyticsApiVersion,
```
Comment (deyaaeldeen) [inline sdk/textanalytics/azure-ai-textanalytics/tests/test_analyze_sentiment.py:19]: are you using a new version of the formatter or something?

File: sdk/textanalytics/azure-ai-textanalytics/tests/test_analyze_sentiment_async.py [hunk 2]
```diff
@@ -17,7 +17,7 @@
     VERSION,
     DetectLanguageInput,
     TextDocumentInput,
-    TextAnalyticsApiVersion
+    TextAnalyticsApiVersion,
```
Comment (deyaaeldeen) [inline sdk/textanalytics/azure-ai-textanalytics/tests/test_analyze_sentiment_async.py:20]: same comment about formatting.

File: sdk/textanalytics/azure-ai-textanalytics/tests/test_detect_language.py [hunk 3]
```diff
@@ -16,7 +16,8 @@
     DetectLanguageInput,
     TextAnalyticsClient,
     DetectLanguageInput,
-    VERSION
+    VERSION,
+    TextAnalyticsApiVersion,
```
Comment (deyaaeldeen) [inline sdk/textanalytics/azure-ai-textanalytics/tests/test_detect_language.py:20]: same comment about formatting.

File: sdk/textanalytics/azure-ai-textanalytics/tests/test_detect_language_async.py [hunk 4]
```diff
@@ -16,7 +16,8 @@
 from azure.ai.textanalytics import (
     VERSION,
     DetectLanguageInput,
-    DetectLanguageInput
+    DetectLanguageInput,
+    TextAnalyticsApiVersion,
```
Comment (deyaaeldeen) [inline sdk/textanalytics/azure-ai-textanalytics/tests/test_detect_language_async.py:20]: same comment about formatting.

File: sdk/textanalytics/azure-ai-textanalytics/tests/test_extract_key_phrases.py [hunk 5]
```diff
@@ -15,7 +15,8 @@
 from azure.ai.textanalytics import (
     TextAnalyticsClient,
     TextDocumentInput,
-    VERSION
+    VERSION,
+    TextAnalyticsApiVersion,
```
Comment (deyaaeldeen) [inline sdk/textanalytics/azure-ai-textanalytics/tests/test_extract_key_phrases.py:19]: same comment about formatting.

File: sdk/textanalytics/azure-ai-textanalytics/tests/test_extract_key_phrases_async.py [hunk 6]
```diff
@@ -16,7 +16,8 @@
 from azure.ai.textanalytics import (
     VERSION,
     DetectLanguageInput,
-    TextDocumentInput
+    TextDocumentInput,
+    TextAnalyticsApiVersion,
```
Comment (deyaaeldeen) [inline sdk/textanalytics/azure-ai-textanalytics/tests/test_extract_key_phrases_async.py:20]: same comment about formatting.

File: sdk/textanalytics/azure-ai-textanalytics/tests/test_recognize_entities.py [hunk 7]
```diff
@@ -15,7 +15,8 @@
 from azure.ai.textanalytics import (
     TextAnalyticsClient,
     TextDocumentInput,
-    VERSION
+    VERSION,
+    TextAnalyticsApiVersion,
```
Comment (deyaaeldeen) [inline sdk/textanalytics/azure-ai-textanalytics/tests/test_recognize_entities.py:19]: same comment about formatting.

File: sdk/textanalytics/azure-ai-textanalytics/tests/test_recognize_entities_async.py [hunk 8]
```diff
@@ -16,7 +16,8 @@
 from azure.ai.textanalytics import (
     VERSION,
     DetectLanguageInput,
-    TextDocumentInput
+    TextDocumentInput,
+    TextAnalyticsApiVersion,
```
Comment (deyaaeldeen) [inline sdk/textanalytics/azure-ai-textanalytics/tests/test_recognize_entities_async.py:20]: same comment about formatting.

File: sdk/textanalytics/azure-ai-textanalytics/tests/test_recognize_linked_entities.py [hunk 9]
```diff
@@ -15,7 +15,8 @@
 from azure.ai.textanalytics import (
     TextAnalyticsClient,
     TextDocumentInput,
-    VERSION
+    VERSION,
+    TextAnalyticsApiVersion,
```
Comment (deyaaeldeen) [inline sdk/textanalytics/azure-ai-textanalytics/tests/test_recognize_linked_entities.py:19]: same comment about formatting.

File: sdk/textanalytics/azure-ai-textanalytics/tests/test_recognize_linked_entities_async.py [hunk 10]
```diff
@@ -16,7 +16,8 @@
 from azure.ai.textanalytics import (
     VERSION,
     DetectLanguageInput,
-    TextDocumentInput
+    TextDocumentInput,
+    TextAnalyticsApiVersion,
```
Comment (deyaaeldeen) [inline sdk/textanalytics/azure-ai-textanalytics/tests/test_recognize_linked_entities_async.py:20]: same comment about formatting.

File: sdk/textanalytics/azure-ai-textanalytics/tests/test_recognize_pii_entities.py [hunk 11]
```diff
@@ -87,11 +87,69 @@ def test_all_successful_passing_text_document_input(self, client):
...
+    @TextAnalyticsClientPreparer()
+    def test_korean_nfd(self, client):
+        result = client.recognize_pii_entities(["아가 SSN: 859-98-0987"])
+        self.assertEqual(result[0].entities[0].offset, 8)
+        self.assertEqual(result[0].entities[0].length, 11)
+
+    @GlobalTextAnalyticsAccountPreparer()
+    @TextAnalyticsClientPreparer()
+    def test_zalgo_text(self, client):
+        result = client.recognize_pii_entities(["ơ̵̧̧̢̳̘̘͕͔͕̭̟̙͎͈̞͔̈̇̒̃͋̇̅͛̋͛̎́͑̄̐̂̎͗͝m̵͍͉̗̄̏͌̂̑̽̕͝͠g̵̢̡̢̡̨̡̧̛͉̞̯̠̤̣͕̟̫̫̼̰͓̦͖̣̣͎̋͒̈́̓̒̈̍̌̓̅͑̒̓̅̅͒̿̏́͗̀̇͛̏̀̈́̀̊̾̀̔͜͠͝ͅ SSN: 859-98-0987"])
```
Comment (deyaaeldeen) [inline sdk/textanalytics/azure-ai-textanalytics/tests/test_recognize_pii_entities.py:147]: interesting, what language is this?

Comment (iscai-msft) [reply sdk/textanalytics/azure-ai-textanalytics/tests/test_recognize_pii_entities.py:147]: i'm not sure, it's called zalta text, it's part of the [list](https://teams.microsoft.com/l/message/19:d6ff4003f5c848a2a2d6dcdfc0ebd497@thread.skype/1582746506991?tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47&groupId=3e17dcb0-4257-4a30-b843-77f47f1d4121&parentMessageId=1582746506991&teamName=Azure%20SDK&channelName=Service%20-%20Text%20Analytics&createdTime=1582746506991) that brian compiled of scenarios to test for encoding

**Comment by iscai-msft** [comment]
@deyaaeldeen in Python it doesn't really matter if we leave a trailing comma during imports, and for arrays taht span multiple lines it's actualy enouraged to leave a trailign comma. Which parameter are you saying we should come up with a better name for?

**Comment by deyaaeldeen** [comment]
@iscai-msft `string-index-type`. I think what it really specifies is which unit should we use to measure the length of the text with so type of index does not feel right to me. Perhaps `string-code-unit` sounds better?

**Comment by iscai-msft** [comment]
@deyaaeldeen this parameter isn't exposed to users at all for python, so naming doesn't matter since it's just for us. You should sync with @willmtemple to see if JS wants to expose string-index-type to users (at least initially)

**Comment by maririos** [comment]
> @deyaaeldeen this parameter isn't exposed to users at all for python, so naming doesn't matter since it's just for us.

I would say it matters for whoever is maintaining the library, so even if we don't expose it, we should consider a name that indicates what the variable is storing in place.
I agree with @deyaaeldeen that `string-code-unit` is better than `string-index-type`

**Comment by deyaaeldeen** [comment]
@iscai-msft I think code readability is also important and coming up with good names is generally a good thing. It is a nitpick though so I am happy with whatever call you make on this.

File: sdk/textanalytics/azure-ai-textanalytics/azure/ai/textanalytics/_text_analytics_client.py [hunk 12]
```diff
@@ -93,6 +93,7 @@ def __init__(self, endpoint, credential, **kwargs):
         )
         self._default_language = kwargs.pop("default_language", "en")
         self._default_country_hint = kwargs.pop("default_country_hint", "US")
+        self._string_index_type = None if kwargs.get("api_version") == "v3.0" else "UnicodeCodePoint"
```
Comment (maririos) [inline sdk/textanalytics/azure-ai-textanalytics/azure/ai/textanalytics/_text_analytics_client.py:96]: would there be any possibility that in the future the user will be able to change this?
My thinking is.. if we should have this as `default_string_index_type`

File: sdk/textanalytics/azure-ai-textanalytics/azure/ai/textanalytics/_text_analytics_client.py [hunk 13]
```diff
@@ -213,6 +214,8 @@ def recognize_entities(  # type: ignore
         docs = _validate_input(documents, "language", language)
         model_version = kwargs.pop("model_version", None)
         show_stats = kwargs.pop("show_stats", False)
+        if self._string_index_type:
```
Comment (maririos) [inline sdk/textanalytics/azure-ai-textanalytics/azure/ai/textanalytics/_text_analytics_client.py:217]: This is me learning Python....
if the user for whatever reason finds out we have this parameter, and decides to pass it by using `kwargs` the library wouldn't read it and will always use the default value for the library, right?

Comment (maririos) [reply sdk/textanalytics/azure-ai-textanalytics/azure/ai/textanalytics/_text_analytics_client.py:217]: same if it is V3.0 of the service which I think this line wouldn't execute as `self._string_index_type` will be set to `None`

Comment (iscai-msft) [reply sdk/textanalytics/azure-ai-textanalytics/azure/ai/textanalytics/_text_analytics_client.py:217]: yeah, we never read `string_index_type` from kwargs from users. It's just if the api version is not v3.0, we add string_index_type == "UnicodeCodePoint" to the kwargs for the actual network call

File: sdk/textanalytics/azure-ai-textanalytics/azure/ai/textanalytics/_text_analytics_client.py (hunk 12 above)
Comment (iscai-msft) [reply sdk/textanalytics/azure-ai-textanalytics/azure/ai/textanalytics/_text_analytics_client.py:96]: this property isn't exposed to users, so I can change this code anytime (it's just an internal property to keep track of the python encoding). If in the future we do decide to expose the param with name `string_index_type` to users, this won't interfere at all

Comment (maririos) [reply sdk/textanalytics/azure-ai-textanalytics/azure/ai/textanalytics/_text_analytics_client.py:96]: I understand. I am wondering if we should append `default` to the name, even internally, for consistency with the other properties and to tell whoever reads/works on this code later, that this is the default value assigned for the client

File: sdk/textanalytics/azure-ai-textanalytics/tests/test_recognize_pii_entities.py [hunk 14]
```diff
@@ -87,11 +87,69 @@ def test_all_successful_passing_text_document_input(self, client):
...
     @GlobalTextAnalyticsAccountPreparer()
     @TextAnalyticsClientPreparer()
-    def test_length_with_emoji(self, client):
+    def test_emoji(self, client):
         result = client.recognize_pii_entities(["👩 SSN: 859-98-0987"])
         self.assertEqual(result[0].entities[0].offset, 7)
         self.assertEqual(result[0].entities[0].length, 11)
 
+    @GlobalTextAnalyticsAccountPreparer()
+    @TextAnalyticsClientPreparer()
```
Comment (maririos) [inline sdk/textanalytics/azure-ai-textanalytics/tests/test_recognize_pii_entities.py:96]: this tests are not specifically for PII. Looks like Python has files per group of tests, so consider moving this to a specific file and test it for different endpoints too

**Comment by maririos** [comment]
With this change, I think it is worth adding in the readme, and in the description of `Length` and `Offset` that the encoding used is `UnicodeCodePoint`

File: sdk/textanalytics/azure-ai-textanalytics/tests/test_recognize_pii_entities.py (hunk 14 above)
Comment (iscai-msft) [reply sdk/textanalytics/azure-ai-textanalytics/tests/test_recognize_pii_entities.py:96]: I'll move this into another test file for offset and length tests

**Comment by iscai-msft** [comment]
@maririos I really don't have strong feelings, so I'll just change the internal name to `string_code_unit`. I don't think the encoding is worth mentioning in the readme, but I'll add it to the docstrings

**Pull request merged**